### ISSUE TITLE
conduit-extra-1.3.0 test suite failure

### ISSUE THREAD

**Issue by peti** [snoyberg/conduit#385]
Citing from https://hydra.nixos.org/build/79515049:

~~~
Failures:

  test/Data/Conduit/Process/TypedSpec.hs:11:3: 
  1) Data.Conduit.Process.Typed cat works
       uncaught exception: ExitCodeException
       Received ExitFailure (-15) when running
       Raw command: cat ChangeLog.md

  To rerun use: --match "/Data.Conduit.Process.Typed/cat works/"

Randomized with seed 390678105

Finished in 13.9591 seconds
132 examples, 1 failure
Test suite test: FAIL
~~~

**Comment by snoyberg** [comment]
Does this happen reliably? Do you have any repro steps? This does not occur
on CI or the Stackage build server.

**Comment by peti** [comment]
Here is another instance of the test suite failure: https://hydra.nixos.org/build/83308531

**Comment by infinity0** [comment]
This failure happens for us about 2/3 of the time when running as part of a nix build.

**Comment by qrilka** [comment]
I've seen the same error with Stack tests, failure doesn't appear to be reproducible

**Comment by qrilka** [comment]
It looks like there's some concurrency issue happening, e.g. with the following stack.yaml https://gist.github.com/qrilka/176a1a1d92d96e10ffa953d3cd36925b and (a bit modified) Stack from master I get this failure quite reliably, at the same time it doesn't happen when I runt tests only for this package or if I limit concurrency to 1 thread with `-j 1`. Not sure how to minimize it further

**Comment by kquick** [comment]
I am seeing this failure as well.

It's worth noting that this test will require several more open file descriptors as well as other resources, so the non-determinism may be related to consumption/limits of these various system resources.

I would also like to suggest that the failure to run `cat` should disable this set of tests, but that it should not be an actual test failure.  This would help for non-POSIX systems where there is no `cat` executable.

**Comment by seanparsons** [comment]
Testing this now it fails consistently on that one test (the others running cat work fine) with nix, but building it with cabal seems to work every time I've tried. This is on NixOS as well, so the two types of build shouldn't be _that_ far apart.

**Comment by cprussin** [comment]
Any word on this issue?  Looks like this is breaking folks from being able to build packages with https://github.com/cprussin/nixjs/issues/3... I could probably figure out a way to overlay this package and disable tests from running or something but if a fix is on the horizon I'll just wait instead...

**Comment by snoyberg** [comment]
I have no reliable repro I can run of this. If someone on Nix wants to try to debug it, I'll be happy to take a PR. I don't understand why a failing test suite is blocking users from using this package (and don't care to argue that point here). I have not encountered that kind of problem with any other build tool.

**Comment by peti** [comment]
@snoyberg,

> I have no reliable repro I can run of this.

Did you see https://github.com/snoyberg/conduit/issues/385#issuecomment-449934958 and https://github.com/snoyberg/conduit/issues/385#issuecomment-450348192? These observations seem to be a pretty good indication that the issue is not limited to (and therefore not caused by) Nix.

**Comment by snoyberg** [comment]
Right after writing the above comment, I had an idea for what might be causing this. I've pushed 20fd6e2204e84ecf9c1ce99ed3778210a12545ff. If someone wants to test and confirm, I'll merge.

**Comment by bbarker** [comment]
I'd be happy to test this, but as I sat down to do it, I realized I wasn't sure the best way to do it in nix - would love to know (cc @peti). I tried something anyway but ran into what lookes like maybe a hackage issue? 

```
$ stack --nix build                                              
Selected mirror https://s3.amazonaws.com/hackage.fpcomplete.com/                                
Downloading root                                                                                
Unable to load cabal files for snapshot                                                         
                                                                                                
----                                                                                            
Deleting cached snapshot file: /home/brandon/.stack/build-plan/lts-13.6.yaml                    
Recommendation: try running again. If this fails again, open an upstream issue at:              
https://github.com/fpco/lts-haskell/issues/new                                                  
----                                                                                            
                                                                                                
Updating package index Hackage (mirrored at https://s3.amazonaws.com/hackage.fpcomplete.com/) ...
/home/brandon/.stack/indices: createDirectory: resource exhausted (No space left on device)   
```

( I tried a few more times, but all within a couple of minutes of each other)

**Comment by cprussin** [comment]
> I don't understand why a failing test suite is blocking users from using this package (and don't care to argue that point here). I have not encountered that kind of problem with any other build tool.

FWIW @snoyberg the reason for this is that the default behavior in nix is to run the test suite for any package that is built.  In many packages specific tests or the whole suite is disabled, but as a whole, they do attempt to run the suite.  I'm not so sure I agree with nix doing this (or disagree, or really have any opinion at all), but hey, it is what it is.

In any case, it would be possible to put in an exception for specific packages that have known broken tests, but obviously I think it's better to report broken tests upstream like in this case so they can be fixed in the packages.

Long story short--it's a bit misleading for me to have called this a blocker--my apologies.

**Comment by bbarker** [comment]
@snoyberg I tested master and then 20fd6e2204e84ecf9c1ce99ed3778210a12545ff, but still get:

```
        +++ OK, passed 100 tests.
      sourceNull
        +++ OK, passed 100 tests.
      sourceNullS
        +++ OK, passed 100 tests.
    
    Failures:
    
      test/Spec.hs:350:5: 
      1) Combinators stdout
           uncaught exception: IOException of type InvalidArgument
           /run/user/1001/capture24349-117: hGetContents: invalid argument (invalid byte sequence)
           (after 6 tests and 3 shrinks)
             ["\1099324"]
    
      To rerun use: --match "/Combinators/stdout/"
    
      test/Spec.hs:354:5: 
      2) Combinators stderr
           uncaught exception: IOException of type InvalidArgument
           /run/user/1001/capture24349-148: hGetContents: invalid argument (invalid byte sequence)
           (after 6 tests and 3 shrinks)
             ["\1099324"]
    
      To rerun use: --match "/Combinators/stderr/"
    
    Randomized with seed 347217270
    
    Finished in 4.5719 seconds
    343 examples, 2 failures

```

**Comment by nh2** [comment]
I have also just observed this (using a version without the patch mentioned).

It happened the first time I built a set of package in parallel. But afterwards I could not reproduce it by re-building the same package 100s of times in a row. Perhaps it only fails/failed when there's some form of load on the system?

**Comment by robx** [comment]
I can confirm that this is fixed by https://github.com/snoyberg/conduit/commit/20fd6e2204e84ecf9c1ce99ed3778210a12545ff

I've just encountered the same bug with version 1.3.0 on guix:

"Data.Conduit.Process.Typed cat works"

was failing reliably (on a quiet but weak linux virtual machine)

After applying the patch, the test passes.
